kafka-client/src/main/java/dev/responsive/kafka/internal/metrics/ResponsiveRestoreListener.java

```diff
@@ -0,0 +1,227 @@
+import static dev.responsive.kafka.internal.metrics.StoreMetrics.RESTORE_LATENCY;
+import static dev.responsive.kafka.internal.metrics.StoreMetrics.RESTORE_LATENCY_AVG;
+import static dev.responsive.kafka.internal.metrics.StoreMetrics.RESTORE_LATENCY_DESCRIPTION;
+import static dev.responsive.kafka.internal.metrics.StoreMetrics.RESTORE_LATENCY_MAX;
+
+import java.io.Closeable;
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.kafka.common.MetricName;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.metrics.Gauge;
+import org.apache.kafka.common.metrics.Sensor;
+import org.apache.kafka.common.metrics.stats.Avg;
+import org.apache.kafka.common.metrics.stats.CumulativeCount;
+import org.apache.kafka.common.metrics.stats.Max;
+import org.apache.kafka.streams.processor.StateRestoreListener;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Note: this is a global object which is shared across all StreamThreads and state stores in
+ * this Streams application.
+ */
+public class ResponsiveRestoreListener implements StateRestoreListener, Closeable {
+  private static final Logger LOG = LoggerFactory.getLogger(ResponsiveRestoreListener.class);
+
+  private final ResponsiveMetrics metrics;
+  private final MetricName numRestoringMetricName;
+  private final Sensor numInterruptedSensor;
+  private final Sensor restoreLatencySensor;
+
+  private final Map<TopicPartition, Long> restoringChangelogToStartMs = new HashMap<>();
+
+  private StateRestoreListener userRestoreListener;
+
+  public ResponsiveRestoreListener(final ResponsiveMetrics metrics) {
+    this.metrics = metrics;
+
+    numRestoringMetricName = metrics.metricName(
+        NUM_RESTORING_CHANGELOGS,
+        NUM_RESTORING_CHANGELOGS_DESCRIPTION,
+        metrics.applicationLevelMetric()
+    );
+    numInterruptedSensor = metrics.addSensor(NUM_INTERRUPTED_CHANGELOGS);
+    restoreLatencySensor = metrics.addSensor(RESTORE_LATENCY);
+
+    metrics.addMetric(
+        numRestoringMetricName,
+        (Gauge<Integer>) (config, now) -> restoringChangelogToStartMs.size()
+    );
+
+    numInterruptedSensor.add(
+        metrics.metricName(
+            NUM_INTERRUPTED_CHANGELOGS,
+            NUM_INTERRUPTED_CHANGELOGS_DESCRIPTION,
+            metrics.applicationLevelMetric()),
+        new CumulativeCount()
+    );
+  }
+
+  public void registerUserRestoreListener(final StateRestoreListener restoreListener) {
+    userRestoreListener = restoreListener;
+  }
+
+  public StateRestoreListener userRestoreListener() {
+    return userRestoreListener;
+  }
+
+  @Override
+  public synchronized void onRestoreStart(
+      final TopicPartition topicPartition,
+      final String storeName,
+      final long startingOffset,
+      final long endingOffset
+  ) {
+    final long startMs = System.currentTimeMillis();
+    restoringChangelogToStartMs.put(topicPartition, startMs);
+
+    restoreLatencySensor.add(
```

Comment:
it feels a bit weird to track this using windowed stats live average/max. You'd see NaN (or maybe 0?) except for the window where restore finished and then you'd see one data point that has the restore latency, and then back to NaN/0. Maybe it would make more sense as a gauge that reflects the current time the store has been restoring, and that returns -1 or 0 when the store is not restoring? That way we can detect when something has been restoring for a long time?

Reply:
Mm yeah I suspected this wasn't quite right but thought it'd be better than nothing...your suggestion makes a lot more sense, will open a PR with this change